include/nbl/builtin/hlsl/bxdf/ndf/ggx.hlsl

```diff
@@ -0,0 +1,190 @@
+            factor *= -scalar_type(2.0) * VdotHLdotH / (VdotH_etaLdotH * VdotH_etaLdotH);
+        }
+        return FNG * factor;
+    }
+
+    scalar_type DG1(scalar_type ndf, scalar_type G1_over_2NdotV)
+    {
+        return ndf * scalar_type(0.5) * G1_over_2NdotV;
+    }
+
+    scalar_type DG1(scalar_type ndf, scalar_type G1_over_2NdotV, bool transmitted, scalar_type VdotH, scalar_type LdotH, scalar_type VdotHLdotH, scalar_type orientedEta, scalar_type reflectance)
+    {
+        scalar_type FN = hlsl::mix(reflectance, scalar_type(1.0) - reflectance, transmitted) * ndf;
+        return FVNDF(FN, G1_over_2NdotV, transmitted, VdotH, LdotH, VdotHLdotH, orientedEta);
+    }
+
+    scalar_type devsh_part(scalar_type NdotX2, scalar_type a2, scalar_type one_minus_a2)
+    {
+        return sqrt(a2 + one_minus_a2 * NdotX2);
+    }
+
+    scalar_type G1_wo_numerator(scalar_type NdotX, scalar_type NdotX2, scalar_type a2, scalar_type one_minus_a2)
+    {
+        return scalar_type(1.0) / (NdotX + devsh_part(NdotX2,a2,one_minus_a2));
+    }
+
+    scalar_type G1_wo_numerator(scalar_type NdotX, scalar_type devsh_part)
+    {
+        return scalar_type(1.0) / (NdotX + devsh_part);
+    }
+
+    scalar_type correlated_wo_numerator(scalar_type a2, scalar_type NdotV, scalar_type NdotV2, scalar_type NdotL, scalar_type NdotL2)
+    {
+        scalar_type one_minus_a2 = scalar_type(1.0) - a2;
+        scalar_type Vterm = NdotL * devsh_part(NdotV2, a2, one_minus_a2);
+        scalar_type Lterm = NdotV * devsh_part(NdotL2, a2, one_minus_a2);
+        return scalar_type(0.5) / (Vterm + Lterm);
+    }
+
+    scalar_type G2_over_G1(scalar_type a2, bool transmitted, scalar_type NdotV, scalar_type NdotV2, scalar_type NdotL, scalar_type NdotL2)
+    {
+        scalar_type G2_over_G1;
+        if (transmitted)
+        {
+            scalar_type one_minus_a2 = scalar_type(1.0) - a2;
+            scalar_type devsh_v = devsh_part(NdotV2, a2, one_minus_a2);
+            scalar_type L_v = scalar_type(0.5) * (devsh_v / NdotV - scalar_type(1.0));
+            scalar_type L_l = scalar_type(0.5) * (devsh_part(NdotL2, a2, one_minus_a2) / NdotL - scalar_type(1.0));
+            G2_over_G1 = hlsl::beta<scalar_type>(scalar_type(1.0) + L_l, scalar_type(1.0) + L_v);
+            G2_over_G1 *= scalar_type(1.0) + L_v;
+        }
+        else
+        {
+            scalar_type one_minus_a2 = scalar_type(1.0) - a2;
+            scalar_type devsh_v = devsh_part(NdotV2, a2, one_minus_a2);
+            G2_over_G1 = NdotL * (devsh_v + NdotV); // alternative `Vterm+NdotL*NdotV /// NdotL*NdotV could come as a parameter
+            G2_over_G1 /= NdotV * devsh_part(NdotL2, a2, one_minus_a2) + NdotL * devsh_v;
+        }
+
+        return G2_over_G1;
+    }
+};
+
+template<typename T>
+NBL_PARTIAL_REQ_TOP(concepts::FloatingPointScalar<T>)
+struct GGX<T,true NBL_PARTIAL_REQ_BOT(concepts::FloatingPointScalar<T>) >
+{
+    using scalar_type = T;
```

Comment:
you can keep things that don't vary with V and L as members, so roughness and things connected to it

`a2`, `one_minus_a2` and so on